cmd/gitserver/server/vcs_syncer.go

```diff
@@ -302,3 +308,141 @@ func (s *PerforceDepotSyncer) RemoteShowCommand(ctx context.Context, remoteURL *
+func (s MavenArtifactSyncer) CloneCommand(ctx context.Context, remoteURL *vcs.URL, tmpPath string) (cmd *exec.Cmd, err error) {
+	repository, groupID, artifactID := decomposeMavenURL(remoteURL)
+	versions, err := coursier.ListVersions(ctx, repository, groupID, artifactID)
+	if err != nil {
+		return nil, err
+	}
+
+	paths, err := coursier.FetchVersions(ctx, repository, groupID, artifactID, versions)
+	if err != nil {
+		return nil, err
+	}
+
+	initCmd := exec.CommandContext(ctx, "git", "init")
+	initCmd.Dir = tmpPath
+	if output, err := runWith(ctx, cmd, false, nil); err != nil {
+		return nil, errors.Wrapf(err, "failed to init git repository with output %q", string(output))
+	}
+
+	return nil, s.commitJars(ctx, GitDir(tmpPath), paths, versions)
+}
+
+var versionPattern = lazyregexp.New(`refs/heads/(.+)$`)
+
+// Fetch tries to fetch updates from the remote to given directory.
+func (s MavenArtifactSyncer) Fetch(ctx context.Context, remoteURL *vcs.URL, dir GitDir) error {
+	repository, groupID, artifactID := decomposeMavenURL(remoteURL)
+	coursierVersions, err := coursier.ListVersions(ctx, repository, groupID, artifactID)
+	if err != nil {
+		return err
+	}
+
+	cmd := exec.CommandContext(ctx, "git", "show-branch")
+	dir.Set(cmd)
+	gitVersionsRaw, err := runWith(ctx, cmd, false, nil)
+	if err != nil {
+		return errors.Wrapf(err, "failed to git show-branch with output %q", string(gitVersionsRaw))
+	}
+	gitVersionsLines := strings.Split(string(gitVersionsRaw), "\n")
+
+	gitVersions := make(map[string]struct{})
+	for _, rawVersion := range gitVersionsLines {
+		gitVersions[versionPattern.FindStringSubmatch(rawVersion)[1]] = struct{}{}
+	}
+	var filteredCoursierVersions []string
+	for _, version := range coursierVersions {
+		if _, exists := gitVersions[version]; !exists {
+			filteredCoursierVersions = append(filteredCoursierVersions, version)
+		}
+	}
+
+	paths, err := coursier.FetchVersions(ctx, repository, groupID, artifactID, filteredCoursierVersions)
+
+	return s.commitJars(ctx, dir, paths, filteredCoursierVersions)
+}
+
+func (s MavenArtifactSyncer) commitJars(ctx context.Context, dir GitDir, paths, versions []string) error {
+	for idx, path := range paths {
+		cmd := exec.CommandContext(ctx, "git", "checkout", "--orphan", versions[idx])
+		dir.Set(cmd)
+		if output, err := runWith(ctx, cmd, false, nil); err != nil {
+			return errors.Wrapf(err, "failed to git checkout with output %q", string(output))
+		}
+
+		cmd = exec.CommandContext(ctx, "git", "reset")
+		dir.Set(cmd)
+		if output, err := runWith(ctx, cmd, false, nil); err != nil {
+			return errors.Wrapf(err, "failed to git reset with output %q", string(output))
+		}
+
+		cmd = exec.CommandContext(ctx, "git", "clean", "-d", "-f")
+		dir.Set(cmd)
+		if output, err := runWith(ctx, cmd, false, nil); err != nil {
+			return errors.Wrapf(err, "failed to git clean with output %q", string(output))
+		}
+
+		cmd = exec.CommandContext(ctx, "unzip", path, "-d", "./")
```

Comment:
Not sure if this is needed, but might not hurt to always overwrite existing files
```suggestion
		cmd = exec.CommandContext(ctx, "unzip", "-o", path, "-d", "./")
```

Reply:
i think i'd actually rather not overwrite: if there's existing files, it means the git clean command failed, which means there could also be files that won't get overwritten but will erroneously be included in the commit